btree/btree.c

```diff
@@ -3679,38 +3694,38 @@ static int64_t btree_get_tick(struct m0_sm_op *smop)
 					return fail(bop, M0_ERR(-EFAULT));
 				}
 				oi->i_used++;
-				return node_get(&oi->i_nop, tree,
-						&child, P_NEXTDOWN);
+				lock_acquired = bop->bo_flags & BOF_LOCKALL;
+				return node_get(&oi->i_nop, tree, &child,
+						lock_acquired, P_NEXTDOWN);
 			} else
 				return P_LOCK;
 		} else {
 			node_op_fini(&oi->i_nop);
 			return fail(bop, oi->i_nop.no_op.o_sm.sm_rc);
 		}
 	case P_LOCK:
-		if (!locked(tree))
-			return lock_op_init(&bop->bo_op, &bop->bo_i->i_nop,
-					    bop->bo_arbor->t_desc, P_CHECK);
-		else
-			return P_CHECK;
+		return lock_op_init(&bop->bo_op, &bop->bo_i->i_nop,
+				    bop->bo_arbor->t_desc, P_CHECK);
 	case P_CHECK:
-		oi->i_trial++;
 		if (!path_check(oi, tree, &bop->bo_rec.r_key.k_cookie)) {
-			if (oi->i_trial == MAX_TRIALS) {
-				if (bop->bo_flags & BOF_LOCKALL)
+			if (oi->i_trial >= MAX_TRIALS) {
+				if (bop->bo_flags & BOF_LOCKALL) {
+					lock_op_unlock(bop->bo_arbor->t_desc);
 					return fail(bop, -ETOOMANYREFS);
-				else
+				} else
 					bop->bo_flags |= BOF_LOCKALL;
-			}
+			} else
+				oi->i_trial++;
```

Comment:
No need of else statement, you can add `oi->i_trial++ just after path_check on line 3711. Likewise in other statemachines.

Reply:
that can be done too.

Reply:
done